python/taichi/ad/_ad.py

```diff
@@ -85,14 +103,133 @@ def __exit__(self, _type, value, tb):
             self.grad()
 
     def insert(self, func, args):
+        # print("insert!")
         self.calls.append((func, args))
 
     def grad(self):
         assert self.entered, "Before evaluating gradients tape must be entered."
         assert not self.gradient_evaluated, "Gradients of grad can be evaluated only once."
         for func, args in reversed(self.calls):
-            func.grad(*args)
+            if hasattr(func, 'grad'):
```

Comment:
Just wondering if there are cases failed without this line?

Reply:
Yes, because now we also record write_float and write_int kernels